src/vec_int128_ppc.h

```diff
@@ -2388,73 +2889,366 @@ vec_mulluq (vui128_t a, vui128_t b)
+  tc = (vui32_t) vec_addcuq ((vui128_t) t_odd, (vui128_t) t);
+  t_odd = (vui32_t) vec_adduqm ((vui128_t) t_odd, (vui128_t) t);
 
-/*! \brief Vector Byte Reverse Quadword.
- *
- *  Return the bytes / octets of a 128-bit vector in reverse order.
- *
- *  |processor|Latency|Throughput|
- *  |--------:|:-----:|:---------|
- *  |power8   | 2-13  | 2 cycle  |
- *  |power9   | 3     | 2/cycle  |
- *
- *  @param vra a 128-bit vector treated a __int128.
- *  @return a 128-bit vector with the bytes in reserve order.
- */
-static inline vui128_t
-vec_revbq (vui128_t vra)
-{
-  vui128_t result;
+  /* rotate right the low 32-bits into tmq */
+  tmq = vec_sld (t_odd, tmq, 12);
+  /* shift the low 128 bits (with carry) of partial product right
+   * 32-bits */
+  t_odd = vec_sld (tc, t_odd, 12);
+  /* add the top 128 bits of even / odd partial products */
+  t = (vui32_t) vec_adduqm ((vui128_t) t_even, (vui128_t) t_odd);
 
-#ifdef _ARCH_PWR9
-#ifndef vec_revb
-  __asm__(
+  tsw = vec_splat ((vui32_t) b, VEC_WE_1);
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+  t_even = (vui32_t) vec_mulouw ((vui32_t) a, tsw);
+  t_odd = (vui32_t) vec_muleuw ((vui32_t) a, tsw);
+#else
+  t_even = (vui32_t)vec_muleuw((vui32_t)a, tsw);
+  t_odd = (vui32_t)vec_mulouw((vui32_t)a, tsw);
+#endif
+  /* Sum the low 128 bits of odd previous partial products */
+  tc = (vui32_t) vec_addcuq ((vui128_t) t_odd, (vui128_t) t);
+  t_odd = (vui32_t) vec_adduqm ((vui128_t) t_odd, (vui128_t) t);
+
+  /* rotate right the low 32-bits into tmq */
+  tmq = vec_sld (t_odd, tmq, 12);
+  /* shift the low 129 bits (with carry) of partial product right
+   * 32-bits */
+  t_odd = vec_sld (tc, t_odd, 12);
+  /* add the top 128 bits of even / odd partial products */
+  t = (vui32_t) vec_adduqm ((vui128_t) t_even, (vui128_t) t_odd);
+
+  tsw = vec_splat ((vui32_t) b, VEC_WE_0);
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+  t_even = (vui32_t) vec_mulouw ((vui32_t) a, tsw);
+  t_odd = (vui32_t) vec_muleuw ((vui32_t) a, tsw);
+#else
+  t_even = (vui32_t)vec_muleuw((vui32_t)a, tsw);
+  t_odd = (vui32_t)vec_mulouw((vui32_t)a, tsw);
+#endif
+  /* Sum the low 128 bits of odd previous partial products,
+   * generate the carry.  */
+  tc = (vui32_t) vec_addcuq ((vui128_t) t_odd, (vui128_t) t);
+  t_odd = (vui32_t) vec_adduqm ((vui128_t) t_odd, (vui128_t) t);
+
+  /* rotate right the low 32-bits into tmq */
+  tmq = vec_sld (t_odd, tmq, 12);
+  /* shift the low 129 bits (with carry) of partial product right
+   * 32-bits */
+  t_odd = vec_sld (tc, t_odd, 12);
+  /* add the top 128 bits of even / odd partial products */
+  t = (vui32_t) vec_adduqm ((vui128_t) t_even, (vui128_t) t_odd);
+#else // _ARCH_PWR7 or earlier and Big Endian only.  */
+  /* We use the Vector Multiple Even/Odd Unsigned haldword to compute
+   * the 128 x 16 partial (144-bit) product of value a with the
+   * halfword splat of b. These instructions (vmuleuh, vmulouh)
+   * product 8 32-bit 16 x 16 partial products where even
```

Comment:
s/product/produce/

Reply:
Ok we  will try this wording:

  /* We use the Vector Multiple Even/Odd Unsigned Word to compute
   * the 128 x 32 partial (160-bit) product of value a with the
   * word splat of b. This produces four 64-bit (32 x 32)
   * partial products in two vector registers. These results
   * are not aligned for summation as is. So the odd result is
   * shifted right 32-bits before it is summed (via Vector Add
   * Unsigned Quadword Modulo) with the the even result.
   * The low order 32-bits, of the 160-bit product
   * is shifted (right) in to a separate vector (tmq).
   *
   * This is repeated for each (low to high order) word of b.
   * After the first (160-bit) partial product, the high 128-bits
   * (t) of the previous partial product is summed with the current
   * odd multiply result, before this sum (including any carry out)
   * is shifted right 32-bits.  Bits shifted out of the of this sum
   * are shifted (32-bits at a time) into the low order 128-bits
   * of the product (tmq). The shifted odd sum is then added to the
   * current even product, After the 4th step this sum is the
   * final high order 128-bits of the quadword product. */

I plan to include this full test in the _ARCH_PWR8 section of vec_muludq.  muhluq and mulluq
Will include a comment referring to this.